src/memo/memo.ml

```diff
@@ -718,6 +659,98 @@ module Call_stack = struct
     Fiber.Var.set call_stack_var stack (fun () -> Implicit_output.forbid f)
 end
 
+let global_dep_dag = Dag.create ()
+
+module Sample_attempt = struct
+  include Sample_attempt0
+
+  (* Add all edges leading from the root of the call stack to [dag_node] to the
+     cycle detection DAG. *)
+  let add_path_to ~dag_node =
+    let+ stack = Call_stack.get_call_stack () in
+    let rec add_path_impl stack dag_node =
+      match stack with
+      | [] -> Ok ()
+      | Stack_frame_with_state.T
+          ({ dag_node = (lazy caller_dag_node); _ } as frame)
+        :: stack -> (
+        let dag_node_id = Dag.node_id dag_node in
+        match Int.Set.mem frame.children_added_to_dag dag_node_id with
+        | true ->
+          (* Here we know that the current [frame] has already been traversed in
+             a previous [add_path_to] call. Therefore, the DAG already contains
+             all the edges that we will discover by continuing the recursive
+             traversal. We might as well stop here and save time. *)
+          Ok ()
+        | false -> (
+          match
+            Dag.add_assuming_missing global_dep_dag caller_dag_node dag_node
+          with
+          | exception Dag.Cycle cycle ->
+            Error (List.map cycle ~f:(fun dag_node -> dag_node.Dag.data))
+          | () -> (
+            let not_traversed_before =
+              Int.Set.is_empty frame.children_added_to_dag
+            in
+            frame.children_added_to_dag <-
+              Int.Set.add frame.children_added_to_dag dag_node_id;
+            match not_traversed_before with
+            | false ->
+              (* Same optimisation as above. *)
+              Ok ()
+            | true -> add_path_impl stack caller_dag_node)))
+    in
+    add_path_impl stack (Lazy.force dag_node)
+
+  let force_and_check_for_cycles once dag_node =
+    let force () =
+      let+ result = Once.force once in
+      Ok result
+    in
+    match Once.forced once with
```

Comment:
This side-effect outside of a fiber looked suspicious to me at first.
In fact if you consider that `Once.force` is using `Fiber.of_thunk`, then there are two possibilities:
1. That `Fiber.of_thunk` is necessary and there's a bug here
2. That `Fiber.of_thunk` is unnecessary and we're relying on some promise of our caller. (every Fiber we return is called exactly once and immediately after we return it, with no context switch?)

I believe (2) may well be true (hopefully `Fiber.of_thunk` in `exec_dep_node` hopefully insulates us from anything weird the user can throw at us), but how about we stop relying on it here, by changing `force` like this so we don't have to read its status twice:

```
val force : 'a t -> on_wait:(unit, 'a) Fiber.t -> 'a Fiber.t
```

(`on_wait` doesn't sound like the best possible name, ideas are welcome)
(I'm surprised our `Once` module lacks a signature btw)

Reply:
Good point! I like the idea of extending `force` to take an additional `on_wait`-like argument.

Reply:
Done. I called the argument `on_blocking_wait` and also moved `Once` to a separate module with proper signature and comments.

Reply:
Is the word "blocking" supposed to have a different meaning from "wait" here? (sounds a bit repetitive if not)

Reply:
Renamed to `on_blocking` after discussing offline.